client/galaxy/scripts/mvc/tool/tool-form-composite.js

```diff
@@ -319,6 +319,19 @@ define([ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view'
             }
         },
 
+        /** Refresh the history after job submission while form is shown */
+        _refreshHistory: function() {
+            var self = this;
+            var history = parent.Galaxy && parent.Galaxy.currHistoryPanel && parent.Galaxy.currHistoryPanel;
+            this._refresh_history && clearTimeout( this._refresh_history );
+            if ( history ) {
+                history.refreshContents();
+                if ( history.model.numOfUnfinishedShownContents() === 0 ) {
+                    this._refresh_history = setTimeout( function() { self._refreshHistory() }, 2000 );
+                }
+            }
+        },
+
         /** Build remaining steps */
         _execute: function() {
```

Comment:
One is a performance consideration - it is much quicker to just persist the request and take care or it a workflow running thread (these can be configured separately than job runner threads but in most cases are just job runner threads). 

But more essentially - only very simple workflows can be scheduled with all jobs in a new state in advance (given the nature of Galaxy's data model anyway). Imagine if at one step a job splits up an input and creates a collection of unknown jobs - and the next step of the workflow is to map over that collection and do some sort of processing over each item. It is then not known how many jobs there will be corresponding to that processing step until that barrier job that does the splitting is complete.

There are other workflow operations currently implemented that prevent jobs from being created until some sort of barrier action is complete - implicit connections between steps (connections not tied to datasets being created - e.g. data managers), feeding the output of a step as a parameter (non-data) to a subsequent step, dataset review steps. Certainly once conditionals are added - they will be a very clear example as well.

For more information see:

 - https://bitbucket.org/galaxy/galaxy-central/pull-requests/561/allow-background-and-plugin-driven/diff
 - https://bitbucket.org/galaxy/galaxy-central/pull-requests/634/allow-tools-to-explicitly-produce-dataset/diff